lockin/src/lib.rs

```diff
@@ -0,0 +1,973 @@
+/// * `valid_tstamps` - Number of valid timestamps.
+/// * `tstamps_mem` - Last 2 recorded timestamps.
+fn record_new_tstamps(
+    ref_tstamps: [u16; TSTAMP_BUFFER_SIZE],
+    valid_tstamps: usize,
+    tstamps_mem: &mut [TimeStamp; 2],
+) {
+    if valid_tstamps > 1 {
+        tstamps_mem[1].new_count(ref_tstamps[valid_tstamps - 2]);
+        tstamps_mem[0].new_count(ref_tstamps[valid_tstamps - 1]);
+    } else if valid_tstamps == 1 {
+        tstamps_mem[1].count = tstamps_mem[0].count;
+        tstamps_mem[1].sequences_old = tstamps_mem[0].sequences_old;
+        tstamps_mem[0].new_count(ref_tstamps[valid_tstamps - 1]);
+    }
+}
+
+/// Map `iq_to_a` and `iq_to_t` to each pair of `i` and `q`.
+fn iq_to_at_map(
+    i: [f32; OUTPUT_BUFFER_SIZE],
+    q: [f32; OUTPUT_BUFFER_SIZE],
+) -> ([f32; OUTPUT_BUFFER_SIZE], [f32; OUTPUT_BUFFER_SIZE]) {
+    let mut a: [f32; OUTPUT_BUFFER_SIZE] = [0.; OUTPUT_BUFFER_SIZE];
+    let mut t: [f32; OUTPUT_BUFFER_SIZE] = [0.; OUTPUT_BUFFER_SIZE];
+    for k in 0..OUTPUT_BUFFER_SIZE {
+        a[k] = iq_to_a(i[k], q[k]);
+        t[k] = iq_to_t(i[k], q[k]);
+    }
+    (a, t)
+}
+
+/// Returns magnitude from in-phase and quadrature signals.
+///
+/// # Arguments
+///
+/// `i` - In-phase signal.
+/// `q` - Quadrature signal.
+fn iq_to_a(i: f32, q: f32) -> f32 {
+    2. * sqrt(pow2(i) + pow2(q))
+}
+
+/// Returns angle from in-phase and quadrature signals.
+///
+/// # Arguments
+///
+/// `i` - In-phase signal.
+/// `q` - Quadrature signal.
+fn iq_to_t(i: f32, q: f32) -> f32 {
+    atan2(q, i)
+}
+
+/// Demodulation phase values corresponding to each ADC sample.
+///
+/// # Arguments
+///
+/// * `first_t` - First timestamp value from the current processing
+/// period. The value provided here is ignored if there were no
+/// timestamps in the current processing period.
+/// * `tstamps_mem` - Recorded TimeStamps.
+/// * `phi` - Reference phase offset.
+/// * `fscale` - Frequency scaling factor for the demodulation signal.
+/// * `tadc` - ADC sampling period.
+///
+/// # Latency (ADC batch size = 16): 3.3us
+///
+/// 2.9us from computing `real_phase`.
+fn adc_phases(
+    first_t: u32,
+    tstamps_mem: &mut [TimeStamp; 2],
+    phi: u32,
+    fscale: u32,
+    tadc: u32,
+) -> [f32; SAMPLE_BUFFER_SIZE] {
+    let overflow_count: u32 = tadc * SAMPLE_BUFFER_SIZE as u32;
+    let tref_count: u32 = tstamps_diff(tstamps_mem, overflow_count);
+    let mut thetas: [f32; SAMPLE_BUFFER_SIZE] = [0.; SAMPLE_BUFFER_SIZE];
+    let mut theta_count: u32;
+
+    // 68ns
+    if tstamps_mem[0].sequences_old == 0 {
+        theta_count = tref_count - first_t;
+    } else {
+        theta_count = tstamps_diff(
+            &[
+                TimeStamp {
+                    count: 0,
+                    sequences_old: 0,
+                },
+                tstamps_mem[0],
+            ],
+            overflow_count,
+        );
+    }
+
+    thetas[0] = real_phase(theta_count, fscale, tref_count, phi);
```

Comment:
I would recommend a different name than `thetas`. Instead, why not make it `sample_phase` or something similar, so it's clear what it's referring to?